docs/errors.md

```diff
@@ -0,0 +1,124 @@
+# Error Handling
+
+Within the Darklang codebase, consistent and clear error handling is paramount. This guide provides a straightforward set of practices to follow when dealing with different types of errors.
+
+The codebase has a handful of different ways to handle errors. In the runtime we have:
+- F# results
+- darklang results (user-facing version)
+- Exception.raiseInternal
+- Type Errors
+- other runtime errors
+
+## Runtime Errors & Type Errors
+
+The primary way of indicating an error in the runtime and during code execution is a
+`RuntimeError`.
+
+### Using RuntimeError
+
+If you need to raise a RuntimeError, call `RT.Error.raiseRTE`. This throws an
+exception with the RuntimeError in it, which is caught at the execution boundary (eg
+`LibExecution.Execution.executeExpr`). At the execution boundaries, the functions
+returns an F# `Result<Dval, RuntimeError>`, which are then converted to Darklang and
+can be stringified using `@darklang.languageTools.runtimeErrors.error.toString`.
+
+A very common RuntimeError to use is a `TypeChecker.Error.ValueNotExpectedType`,
+which indicates that we have the wrong type. There are helper functions like
+`TypeChecker.raiseValueNotExpectedType` and
+`TypeChecker.raiseFnValResultNotExpectedType` that can make this easier (the latter
+for checking return values from first-class functions called from builtin functions).
+
+We prefer to raise `RuntimeErrorException`s using `raiseRTE` or something similar, vs
+returning F# `Result<_, RuntimeError>`s.
+
+Never catch a `RuntimeError` except in `LibExecution.Execution`. That's not how
+they're intended to be used.
+
+### Getting rid of string-based RuntimeErrors
+
+Many `RuntimeError`s in the codebase at the moment use `string`s, usually via
+`raiseString` or `Runtime.oldError`. We really really want to move away from that.
+Prefer to make a new user-facing `RuntimeError` for your module, or add an extra
+variant to the existing Error enum in a module.
+
+### DvalSource, Location, and raiseUntargetedRTE
+
+We're trying to get to a place where we always include the `DvalSource` of an RuntimeError, but often it can be hard. We plan to ensure we pass around enough information so that RuntimeErrors always have a `DvalSource` - without which we do not have enough
```

Comment:
```suggestion
We're trying to get to a place where we always include the `DvalSource` of an
RuntimeError, but often it can be hard. We plan to ensure we pass around enough
information so that RuntimeErrors always have a `DvalSource` - without which we do
not have enough to quickly find what code caused the error.
```